### PR TITLE
Add iterator close support for for-of statement

### PR DESCRIPTION
**Pull request by rerobika**
JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: tests/jerry/es2015/for-of.js [hunk 1]
```diff
@@ -119,3 +119,109 @@ var i = 0;
 for (var a of obj) {
   assert (a === i++);
 }
+
+function __createIterableObject(arr, methods) {
```
Comment (zherczeg) [inline tests/jerry/es2015/for-of.js]: Why two underscore is needed?

File: tests/jerry/es2015/for-of.js [hunk 2]
```diff
@@ -119,3 +119,109 @@ var i = 0;
...
+  methods = methods || {};
```
Comment (zherczeg) [inline tests/jerry/es2015/for-of.js]: You can use a default parameter

File: jerry-core/vm/vm-stack.c [hunk 3]
```diff
@@ -79,8 +80,11 @@ vm_stack_context_abort (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
 #if ENABLED (JERRY_ES2015)
     case VM_CONTEXT_FOR_OF:
     {
+      ecma_value_t iterator = vm_stack_top_p[-3];
       ecma_free_value (vm_stack_top_p[-2]);
-      ecma_free_value (vm_stack_top_p[-3]);
+
+      ecma_op_iterator_close (iterator);
```
Comment (zherczeg) [inline jerry-core/vm/vm-stack.c]: What are the possible side effects?

File: tests/jerry/es2015/for-of.js (hunk 2 above)
Comment (rerobika) [reply tests/jerry/es2015/for-of.js]: This method comes from the `kangax` test suite, therefore I left the original source.

File: jerry-core/vm/vm-stack.c (hunk 3 above)
Comment (rerobika) [reply jerry-core/vm/vm-stack.c]: This operation may changes the `JERRY_CONTEXT (error_value)` and may applies the `ECMA_STATUS_EXCEPTION` status flags. At this point there is no way to handle exception, so the caller's responsibility to check whether the context abort raised error. See [the example](https://github.com/jerryscript-project/jerryscript/pull/3401/files#diff-4eae7d69004d99f61986f6e351bda007R3632) here.

File: tests/jerry/es2015/for-of.js (hunk 2 above)
Comment (zherczeg) [reply tests/jerry/es2015/for-of.js]: What about the license?

File: jerry-core/vm/vm-stack.c (hunk 3 above)
Comment (zherczeg) [reply jerry-core/vm/vm-stack.c]: What happens if multiple iterators are closed? Nested `for-if`-s. Are they working correctly if their closes throws errors?

Comment (rerobika) [reply jerry-core/vm/vm-stack.c]: The iterator close operation may raises error but it never overwrites the original error which caused the context abort. Each context abort is called, see the example [here](https://github.com/jerryscript-project/jerryscript/pull/3401/files#diff-bf7b2eb18cb71828704e5cc5bad6a567R211).

File: tests/jerry/es2015/for-of.js (hunk 2 above)
Comment (rerobika) [reply tests/jerry/es2015/for-of.js]: [These methods](https://github.com/kangax/compat-table/blob/gh-pages/node.js#L29) only helper functions to run the test suite but useful for testing purposes. I see no license marker from the origin of the file.

Comment (akosthekiss) [reply tests/jerry/es2015/for-of.js]: Beware! Here be dragons! The fact that the file has no license header doesn't mean that it is not covered by a license. (If it had no license at all, that would be the most dangerous situation, actually. That would disallow its reuse anywhere!)

Fortunately, the project repo does contain a [license](https://github.com/kangax/compat-table/blob/gh-pages/LICENSE), which covers all files, including the source reused here. One small problem remains: the git logs claim that this is the MIT license, but actually it isn't: it is some kind of MIT--. The MIT license would contain the condition "The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software", but that's missing from this license. 

Nevertheless, I think it is safe to assume the unmodified MIT license as that's more strict. But then, we should replicate the MIT license in our code. There is example for that, e.g., in jerry-libm.

Some background info on MIT license:
- https://opensource.org/licenses/MIT
- https://tldrlegal.com/license/mit-license

Comment (rerobika) [reply tests/jerry/es2015/for-of.js]: @akosthekiss Thanks for the clarification, I've missed that part.

**Comment by rerobika** [comment]
@zherczeg The patch is updated, please take an other look at it.

File: jerry-core/vm/vm-stack.h [hunk 4]
```diff
@@ -44,13 +44,20 @@
...
  */
 #define VM_CONTEXT_HAS_LEX_ENV 0x20
 
+
+/**
+ * This flag is set if the iterator close operation should be invoked during a for-of context break.
+ */
+#define VM_CONTEXT_CLOSE_ITERATOR 0x40
+
+
```
Comment (zherczeg) [inline jerry-core/vm/vm-stack.h]: One newline.

File: jerry-core/vm/vm-stack.h [hunk 5]
```diff
@@ -28,13 +28,13 @@
 /**
  * Create context on the vm stack.
  */
-#define VM_CREATE_CONTEXT(type, end_offset) ((ecma_value_t) ((type) | ((end_offset) << 6)))
+#define VM_CREATE_CONTEXT(type, end_offset) ((ecma_value_t) ((type) | ((end_offset) << 8)))
```
Comment (zherczeg) [inline jerry-core/vm/vm-stack.h]: Seems 7 is enough. I would not increase it to 8, because it might be misleading.

**Comment by rerobika** [comment]
@zherczeg Thanks for the review, I've updated the patch.

File: tests/jerry/es2015/for-of-iterator-close.js [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  iterable['return'] = methods['return'];
+  iterable['throw'] = methods['throw'];
+
+  return iterable;
+};
+
+function close1() {
+  var closed = false;
+  var iter = createIterable([1, 2, 3], {
+    'return': function(){ closed = true; return {}; }
```
Comment (zherczeg) [inline tests/jerry/es2015/for-of-iterator-close.js]: I would put space after `function()` here, and below.

Comment (rerobika) [reply tests/jerry/es2015/for-of-iterator-close.js]: Sure! Resolved.

**Pull request merged**